File: my-solutions/0907-koko-eating-bananas/solution.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        minK, maxK = 1, max(piles)

        def numHours(k):
            return sum([int(math.ceil(x / k)) for x in piles])


        while(minK <= maxK):
            midK = int((minK + maxK) / 2)
            # print(minK, midK, maxK, numHours(midK))
            if numHours(midK) > h:
                minK = midK + 1
            else:
                maxK = midK - 1

        return minK

File: my-solutions/0907-koko-eating-bananas/solution.cpp
#include<math.h>

class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int l = 0, r = *std::max_element(piles.begin(), piles.end());
        int res = r;
        while(l <= r) {
            int k = (l + r) / 2;
            if(k == 0) { // if k has reduced to 0 it will cause a div by 0 error.
                return res;
            }
            int time = 0;

            // calculate total time taken to finish all piles
            for(auto i = piles.begin(); i != piles.end(); i++) {
                //cout << ceil(*i/k) << endl;
                time += ceil(double(*i)/double(k));
            }

            printf("%d, %d, %d, %d", l, k, r, time); cout << endl;
            if(time > h) {
                l = k + 1;
            } else {
                res = std::min(res, k);
                r = k - 1;
            }
        }
        return res;
    }
};
